File: leetcode-python-go/1326. 灌溉花园的最少水龙头数目/main.py
from typing import List
from numpy import inf

n = 5
ranges = [3,4,1,1,0,0]

def minTaps(n: int, ranges: List[int]) -> int:
    intervals = []
    for i, r in enumerate(ranges):
        start = max(0, i - r)
        end = min(n, i + r)
        intervals.append((start, end))
    intervals.sort()

    dp = [inf] * (n + 1)
    dp[0] = 0
    for start, end in intervals:
        if dp[start] == inf:
            return -1
        for j in range(start, end + 1):
            dp[j] = min(dp[j], dp[start] + 1)
    return dp[n]

print(minTaps(n, ranges))

File: leetcode-python-go/1326. 灌溉花园的最少水龙头数目/main.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func minTaps(n int, ranges []int) int {
	intervals := [][2]int{}
	for i, r := range ranges {
		start := max(0, i-r)
		end := min(n, i+r)
		intervals = append(intervals, [2]int{start, end})
	}
	sort.Slice(intervals, func(i, j int) bool {
		a, b := intervals[i], intervals[j]
		return a[0] < b[0]
	})

	dp := make([]int, n+1)
	for i := range dp {
		dp[i] = math.MaxInt
	}
	dp[0] = 0
	for _, p := range intervals {
		start, end := p[0], p[1]
		if dp[start] == math.MaxInt {
			return -1
		}
		for j := start; j <= end; j++ {
			dp[j] = min(dp[j], dp[start]+1)
		}
	}
	return dp[n]
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func max(a, b int) int {
	if b > a {
		return b
	}
	return a
}

func main() {
	n := 5
	ranges := []int{3, 4, 1, 1, 0, 0}
	results := minTaps(n, ranges)
	fmt.Println(results)
}
